### DISCUSSION TITLE
Java integration impossible

### DISCUSSION THREAD

**Discussion by elbosso** [ethersync/ethersync#220] [Editor plugins]
I am sorry to say that i have to retract my experiment to integrate the protocol with java (swing) as the only events one gets when characters are deleted is the information *after* the edit happened and then it is only of the starting position and the number of removed characters. It is impossible to know, how many lines that removal spanned without keeping a shadow document around and that seems preposterous.

So, in conclusion: I don't think that the current protocol is really very well portable but i also can not estimate if a protocol change towards the java model (ignore line break and treat editor content just as a stream of characters) is feasible to reach the goal of this project.

I will archive the project i set up for integrating the protocol with java and stop working on it now.

PS: After having experimented a bit with it, I don`t think that the concept of "lines" is really good because different Systems might be having different notions about what a line break really is.

**Comment by blinry** [comment]
Hey! :) Thanks for voicing your problems so clearly, that helps us to understand the tradeoffs that are involved!

**Delete operations & shadow document**

We faced the same problem in the Neovim and VS Code plugins: In both cases, at the time of a deletion event, the "old content" is no longer accessible. We solved it the same way you described: Keeping a cache of the old state, so that we can look up how many Unicode characters were deleted (in the VS Code case) or what precisely the differing patch is (in the Neovim case).

So, the solutions I see:

- Indeed keep a content cache for the open files.
	- Good, because it keeps the protocol simple.
	- Neutral, because we've found it relatively quick and easy to implement.
	- Bad, because it has a performance and memory cost, especially for big files. (Could be improved using rope datastructures, but that's more difficult to implement.)
- Adapt the protocol so that it can work with different types of patch formats: VS Code works with a UTF-16 character count, for example (instead of UTF-32, which is what the protocol currently uses). And you might need a "offset to the beginning of the file"-model? (Do you happen to know how Swing counts positions? Is it in bytes/UTF-8/UTF-16/UTF-32?)
	- Good, because it is less work for the plugins.
	- Bad, because it adds new complexity to the protocol.

**Line-oriented positions vs offset-oriented positions**

So far, we've looked at our project from the perspective of providing support to "big, established" text editors. It seems to me that it's pretty common to have a column/line-oriented position scheme there. For example, IntelliJ has the concept of the [logical caret position](https://plugins.jetbrains.com/docs/intellij/coordinates-system.html#caret-logical-position), which is the zero-based line & column position. I think editors mostly do this, because being able to make edits in such a data structure is more performant compared to a "string that contains the entire document".

However, in the application you were trying to build, I guess your text boxes don't have such positions. Again, this could be solved on the client-side by looking up offsets from the beginning of the file in a cache, but I agree that it's a bit annoying. So, the same options as above seem to apply.

Let us know your thoughts, and what other pros and cons you see to the two options! (Or if you maybe see a third option?)
